### PR TITLE
URLSession's delegateQueue should be serial as per the spec.

### PR DESCRIPTION
**Pull request by mamabusi**
Refer to the API document that mentions URLSession's delegateQueue should be serial:
https://developer.apple.com/reference/foundation/urlsession/1411597-init

### CODE REVIEW

File: Foundation/NSURLSession/NSURLSession.swift [hunk 1]
```diff
@@ -226,6 +227,7 @@ open class URLSession : NSObject {
         self.workQueue = DispatchQueue(label: "URLSession<\(identifier)>")
         self.taskAttributesIsolation = DispatchQueue(label: "URLSession<\(identifier)>.taskAttributes", attributes: DispatchQueue.Attributes.concurrent)
         self.delegateQueue = queue ?? OperationQueue()
+        self.delegateQueue.maxConcurrentOperationCount = 1
```
Comment (mamabusi) [inline Foundation/NSURLSession/NSURLSession.swift]: The spec. says "The queue should be a serial queue, in order to ensure the correct ordering of callbacks". But yes  it is not mandatory and user can send in a concurrent queue as well. Tested with a small test-case on Darwin and I see that user's OperationQueue is not converted to serial if it isn't already. Thanks @ianpartridge for pointing this out. 
Modified the changes for the same.

**Comment by pushkarnk** [comment]
@swift-ci test

**Comment by pushkarnk** [comment]
@mamabusi I'd suggest we have a comment making it clear that we will NOT make a user supplied OperationQueue serial.

File: Foundation/NSURLSession/NSURLSession.swift [hunk 2]
```diff
@@ -225,7 +226,9 @@ open class URLSession : NSObject {
         identifier = nextSessionIdentifier()
         self.workQueue = DispatchQueue(label: "URLSession<\(identifier)>")
         self.taskAttributesIsolation = DispatchQueue(label: "URLSession<\(identifier)>.taskAttributes", attributes: DispatchQueue.Attributes.concurrent)
-        self.delegateQueue = queue ?? OperationQueue()
+        let operationQueue = OperationQueue()
```
Comment (ianpartridge) [inline Foundation/NSURLSession/NSURLSession.swift]: We shouldn't create an `OperationQueue` unless we need to. This creates one in all cases.  How about this:

```swift
if queue != nil {
    self.delegateQueue = queue
} else {
    self.delegateQueue = OperationQueue()
    self.delegateQueue.maxConcurrentOperationCount = 1
}
```

Comment (mamabusi) [inline Foundation/NSURLSession/NSURLSession.swift]: Agreed!

**Comment by mamabusi** [comment]
@pushkarnk @ianpartridge Review changes and suggestions are addressed. Thanks!

**Comment by ianpartridge** [comment]
Thanks @mamabusi. @pushkarnk - can you re-run the CI?

**Comment by pushkarnk** [comment]
Looks good to me. Thanks!

**Comment by pushkarnk** [comment]
Just for the records, here's what @ianpartridge found:  https://www.malhal.com/2017/01/07/nsoperationqueue-maxconcurrentoperationcount-1-serial-queue-problem/  (though it is unlikely that a user will supply a delegate method that will block indefinitely)

**Comment by pushkarnk** [comment]
@swift-ci test

**Comment by pushkarnk** [comment]
@swift-ci test

**Comment by pushkarnk** [comment]
@swift-ci test and merge

**Comment by pushkarnk** [comment]
@swift-ci test and merge

**Pull request merged**